### PR TITLE
[MXNET-105] Fix CuDNN performance after code refactor

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
This PR tries to fix the performance degradation report in https://github.com/apache/incubator-mxnet/issues/9874

We observed about 4% performance decrease. There are multiple factors that cause the performance decrease.
* The first one is that the refactored code passes more arrays to the backward of BatchNorm.
* The second one is that the refactored code needs to reinitialize the CuDNN states in every forward and backward.
* The third one is that the refactor code leads to more memory allocation (e.g., creating std::vector) in forward and backward. (However, the test shows that this doesn't cause much performance decrease.)

This PR tries to reduce these overhead.

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: src/operator/nn/batch_norm.cc [hunk 1]
```diff
@@ -486,6 +484,45 @@ static inline bool backward_BatchNormStorageType(const nnvm::NodeAttrs &attrs,
                              dispatch_mode, wanted_mode);
 }
 
+std::vector<nnvm::NodeEntry> BatchNormGrad(const nnvm::NodePtr& n,
+                                           const std::vector<nnvm::NodeEntry>& ograds) {
+  std::vector<nnvm::NodeEntry> out_data(n->num_outputs());
+  for (uint32_t i = 0; i < out_data.size(); ++i) {
+    out_data[i] = nnvm::NodeEntry{n, i, 0};
+  }
+  std::vector<nnvm::NodeEntry> heads;
```
Comment (cjolivier01) [inline src/operator/nn/batch_norm.cc]: Please use reserve()

File: src/operator/nn/batch_norm.cc [hunk 2]
```diff
@@ -486,6 +484,45 @@ static inline bool backward_BatchNormStorageType(const nnvm::NodeAttrs &attrs,
...
+  heads.push_back(ograds[0]);
+  heads.push_back(out_data[batchnorm::kMean]);
+  heads.push_back(out_data[batchnorm::kVar]);
+  heads.push_back(n->inputs[batchnorm::kData]);
+  heads.push_back(n->inputs[batchnorm::kGamma]);
+
+  // add all the auxiliary data
+  //for (uint32_t i = 0; i < prop.aux_states.size(); ++i) {
+  //  inputs.emplace_back(ptr->inputs[i + prop.arguments.size()]);
+  //}
```
Comment (cjolivier01) [inline src/operator/nn/batch_norm.cc]: ?

File: src/operator/nn/batch_norm.cu [hunk 3]
```diff
@@ -690,12 +690,16 @@ void BatchNormGradCompute<gpu>(const nnvm::NodeAttrs& attrs,
...
-  std::vector<TBlob> in_data(inputs.begin() + 3, inputs.begin() + 6);
-  std::vector<TBlob> aux_states(inputs.begin() + 6, inputs.begin() + 8);
-  std::vector<TBlob> out_data(inputs.begin() + 8, inputs.end());
+  std::vector<TBlob> out_data(3);
+  out_data[batchnorm::kMean] = inputs[1];
+  out_data[batchnorm::kVar] = inputs[2];
+  std::vector<TBlob> in_data(3);
+  in_data[batchnorm::kData] = inputs[3];
+  in_data[batchnorm::kGamma] = inputs[4];
+  std::vector<TBlob> aux_states(2);
```
Comment (cjolivier01) [inline src/operator/nn/batch_norm.cu]: What happens to aux states (running mean and variance)?

**Comment by cjolivier01** [comment]
This needs a JIRA ticket

File: src/operator/nn/batch_norm.cc [hunk 4]
```diff
@@ -424,13 +424,19 @@ void BatchNormGradComputeExCPU(const nnvm::NodeAttrs &attrs,
...
   if (SupportMKLDNNBN(inputs[0], param)
       && (inputs[in_data_start].IsMKLDNNData() || inputs[0].IsMKLDNNData())) {
-    std::vector<NDArray> out_grad(inputs.begin(), inputs.begin() + num_out_grads);
-    std::vector<NDArray> in_data(inputs.begin() + in_data_start,
-                                 inputs.begin() + aux_states_start);
-    std::vector<NDArray> aux_states(inputs.begin() + aux_states_start,
-                                    inputs.begin() + out_data_start);
-    std::vector<NDArray> out_data(inputs.begin() + out_data_start, inputs.end());
-    std::vector<NDArray> in_grad(outputs.begin(), outputs.begin() + 3);
+    static thread_local std::vector<NDArray> out_grad(1);
```
Comment (cjolivier01) [inline src/operator/nn/batch_norm.cc]: How does thread_local help here?

Comment (cjolivier01) [reply src/operator/nn/batch_norm.cc]: Won't these hold a reference to the NDARray's Chunk data indefinitely?

File: src/operator/nn/batch_norm.cc (hunk 1 above)
Comment (zheng-da) [reply src/operator/nn/batch_norm.cc]: This code runs to build the computation graph. It only runs once. Do we still need to call reserve()?

File: src/operator/nn/batch_norm.cc (hunk 4 above)
Comment (zheng-da) [reply src/operator/nn/batch_norm.cc]: Here I'm trying to avoid memory allocation for std::vector.

But you are right. It potentially causes mem leak.

File: src/operator/nn/batch_norm.cc (hunk 1 above)
Comment (cjolivier01) [reply src/operator/nn/batch_norm.cc]: yes, please

File: src/operator/nn/batch_norm.cu (hunk 3 above)
Comment (zheng-da) [reply src/operator/nn/batch_norm.cu]: CuDNN version doesn't need aux_states. CUDA version does. So aux_states is set properly to run CUDA code.

Comment (cjolivier01) [reply src/operator/nn/batch_norm.cu]: So, this fix only improves CUDNN operator?  Wouldn't we expect the other non-CUDNN operators to also be slower by the same amount?

File: src/operator/nn/batch_norm-inl.h [hunk 5]
```diff
@@ -261,19 +261,21 @@ void BatchNormGradCompute(const nnvm::NodeAttrs& attrs,
...
-  int aux_states_start = in_data_start + batchnorm::kInMovingMean;
-  int out_data_start = in_data_start + batchnorm::kInMovingVar + 1;
-  std::vector<TBlob> out_grad(inputs.begin(), inputs.begin() + num_out_grads);
-  std::vector<TBlob> in_data(inputs.begin() + in_data_start,
-                             inputs.begin() + aux_states_start);
-  std::vector<TBlob> aux_states(inputs.begin() + aux_states_start,
-                                inputs.begin() + out_data_start);
-  std::vector<TBlob> out_data(inputs.begin() + out_data_start, inputs.end());
-  std::vector<TBlob> in_grad(outputs.begin(), outputs.begin() + 3);
+  static thread_local std::vector<TBlob> out_grad(1);
```
Comment (piiswrong) [inline src/operator/nn/batch_norm-inl.h]: This is probably too many thread_local.
Why are we copying the vectors in the first place?

Comment (zheng-da) [reply src/operator/nn/batch_norm-inl.h]: good question. I'll do that instead.

**Comment by zheng-da** [comment]
opts | no opts | original
-- | -- | --
5200.7825 | 5033.978 | 5251.894
-- | -- | --
43.4050498 | 64.0531217 | 62.5315905

**Comment by piiswrong** [comment]
@cjolivier01

File: src/operator/nn/batch_norm.cu [hunk 6]
```diff
@@ -705,19 +700,18 @@ void BatchNormGradCompute<gpu>(const nnvm::NodeAttrs& attrs,
...
   } else {
     MSHADOW_REAL_TYPE_SWITCH_EX(dtype, DType, AccReal, {
-      BatchNormBackward<gpu, DType, AccReal>(ctx, param, out_grad,
-          in_data, out_data, req, in_grad, aux_states);
+      BatchNormBackward<gpu, DType, AccReal>(ctx, param, inputs, req, outputs);
     })
   }
 #else
+  aux_states[batchnorm::kMovingMean] = inputs[6];
+  aux_states[batchnorm::kMovingVar] = inputs[7];
```
Comment (TaoLv) [inline src/operator/nn/batch_norm.cu]: @zheng-da aux_states is not defined if USE_CUDNN is not enabled. @marcoabreu seems there is no pure cuda ci environment which is not built with cudnn.

Comment (zheng-da) [reply src/operator/nn/batch_norm.cu]: i see. i'll update it.

Comment (zheng-da) [reply src/operator/nn/batch_norm.cu]: agree. @marcoabreu could you add a CI only with CUDA?

Comment (marcoabreu) [reply src/operator/nn/batch_norm.cu]: Compilation only or do we need tests as well?

Comment (zheng-da) [reply src/operator/nn/batch_norm.cu]: I think it's better to run the code at least once. We probably don't need to try both Python2 and Python3, something like that.

Comment (marcoabreu) [reply src/operator/nn/batch_norm.cu]: Done: https://github.com/apache/incubator-mxnet/pull/10281
